### PR TITLE
Sql local

### PR DESCRIPTION
**Pull request by Ericwimsatt**

### CODE REVIEW

File: SqlHandlers.py [hunk 1]
```diff
@@ -1,18 +1,15 @@
...
 from datetime import datetime
 
 
-class LambdaHandler:
+class SqlHandler(DbInterface):
     def __init__(self, host, user, passwd, db):
         # Database Connection
         self.connection = pymysql.connect(host=host, user=user, passwd=passwd, db=db)
 
-    def lambda_handler(self, event=None, context=None):
```
Comment (Ericwimsatt) [inline SqlHandlers.py:13]: This wasn't used and looked safe to delete

File: PandaDbHandlers.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import pandas as pd
+from datetime import datetime
+
+from dbInterface import DbInterface
+
+
+class PandasDbHandler(DbInterface):
```
Comment (Ericwimsatt) [inline PandaDbHandlers.py:7]: This was largely generated with chatGPT; 
Replaces the SQL db with 3 pandas DF

File: main.py [hunk 3]
```diff
@@ -1,139 +1,147 @@
...
-
-endpoint = os.environ.get("ENDPOINT")
-username = os.environ.get("USERNAME")
-password = os.environ.get("PASSWORD")
-database_name = os.environ.get("DB_NAME")
-
-# LambdaHandler API to interact with MySQL database
-import LambdaHandlers
-
-api = LambdaHandlers.LambdaHandler(
```
Comment (Ericwimsatt) [inline main.py:50]: THis stuff isn't really deleted, just moved to a function so I could run it easier from PYCharm

File: main.py [hunk 4]
```diff
@@ -1,139 +1,147 @@
...
+    #         os.path.exists(datahunt_dir)
+    #         and os.path.exists(schema_dir)
+    #         and os.path.exists(results_dir)
+    #     ):
+    #         raise errors.InvalidDataDirectoryError
+    # except errors.InvalidDataDirectoryError:
+    #     print("One of the required data directories does not exist.")
+    #     sys.exit()
+
+    if offlineMode:
```
Comment (Ericwimsatt) [inline main.py:36]: This change is for supporting running w/o sql DB

File: UMutils.py [hunk 5]
```diff
@@ -246,24 +246,17 @@ def logistic(x, k, offset):
...
-    if user_id not in ucs["uuid"].values:  # if this is the user's first task
-        cur_ucs = 0.5
-        var_scores = np.var(task_scores.iloc[-n:])
-        c = logistic(var_scores / (np.log(num_task_scores + 1) / (np.log(a))), 10, 0.2)
-        new_ucs = cur_ucs * (1 - c) + (c) * last_task_score
-        query = "INSERT INTO `ucs` (`uuid`, `score`) VALUES (%s, %s)"
-        cursor.execute(query, (user_id, new_ucs))
-        client.connection.commit()
-    else:
+    cur_ucs = -1
```
Comment (Ericwimsatt) [inline UMutils.py:249]: This is cleaning up duplicate code and moving more into the SQL Handler so the pandas replacement could work better

File: UMutils.py [hunk 6]
```diff
@@ -453,11 +443,13 @@ def scoring_goldstandard(row):
         file = file[file.answer_uuid.str.len() > 3]
 
         # these are the only relevant columns for scoring for now, notice highlight data is not included here
-        cols = ["answer_uuid", "question_Number", "agreed_Answer"]
+        cols = ["answer_uuid", "question_Number", "agreed_Answer", "source_task_uuid"]
```
Comment (Ericwimsatt) [inline UMutils.py:446]: I think this is a bug now, selecting the same answer in different tasks should be treated as two different questions?

File: UMutils.py [hunk 7]
```diff
@@ -65,7 +66,7 @@ def get_module_files_mapping(datahunt, iaa, schema, goldstandard):
             if modulename in schema_file:
                 schema_path = os.path.join(schema, schema_file)
                 mapping[modulename]["Schema"] = schema_path
-        if os.path.exists(goldstandard):
+        if goldstandard is not None and os.path.exists(goldstandard):
```
Comment (Ericwimsatt) [inline UMutils.py:69]: this fixes a path error I was getting

File: main.py [hunk 8]
```diff
@@ -1,139 +1,147 @@
...
-        iaa=iaa_dir,
-        schema=schema_dir,
-        goldstandard=goldstandard_dir,
-    )
-else:
-    module_filemap = UMutils.get_module_files_mapping(
-        datahunt=datahunt_dir, iaa=iaa_dir, schema=schema_dir, goldstandard=None
-    )
-
-if os.path.exists(goldstandard_dir):
```
Comment (Ericwimsatt) [inline main.py:74]: I was getting errors from this line when goldstandard was empty

File: UMutils.py (hunk 6 above)
Comment (Ericwimsatt) [reply UMutils.py:446]: Notably this breaks the scoring that depends on the grouping.

Comment (sushantvema) [reply UMutils.py:446]: I don't fully understand the bug yet, will have to look into it more. Otherwise the other organizational changes look good.

**Review by sushantvema — APPROVED** [review]
Organizational modifications and offline mode features look good. Have to investigate the conditional logic for gold standard data and grouping the scored questions. I don't think that anything will break from this merge.

**Pull request merged**